tests/integration/pilot/bindpodip/bindpodip_test.go

```diff
@@ -0,0 +1,231 @@
+	"istio.io/istio/pkg/test/framework/label"
+	"istio.io/istio/pkg/test/framework/resource"
+	"istio.io/istio/pkg/test/framework/resource/environment"
+	"istio.io/istio/pkg/test/util/retry"
+)
+
+var (
+	i istio.Instance
+	g galley.Instance
+	p pilot.Instance
+)
+
+// TestMain defines the entrypoint for pilot tests using a standard Istio installation.
+// If a test requires a custom install it should go into its own package, otherwise it should go
+// here to reuse a single install across tests.
+func TestMain(m *testing.M) {
+	framework.
+		NewSuite("bindpodip_test", m).
+		Label(label.CustomSetup).
+		RequireEnvironment(environment.Kube).
+		SetupOnEnv(environment.Kube, istio.Setup(&i, setupConfig)).
+		Setup(func(ctx resource.Context) (err error) {
+			if g, err = galley.New(ctx, galley.Config{}); err != nil {
+				return err
+			}
+			return nil
+		}).
+		Setup(func(ctx resource.Context) (err error) {
+			if p, err = pilot.New(ctx, pilot.Config{
+				Galley: g,
+			}); err != nil {
+				return err
+			}
+			return nil
+		}).
+		Run()
+}
+
+func TestBindPodIPPorts(t *testing.T) {
+	framework.
+		NewTest(t).
+		RequiresEnvironment(environment.Kube).
+		Run(func(ctx framework.TestContext) {
+			ns := namespace.NewOrFail(t, ctx, namespace.Config{
+				Prefix: "bindpodip",
+				Inject: true,
+			})
+
+			var client echo.Instance
+			var instance echo.Instance
+			echoboot.NewBuilderOrFail(t, ctx).
+				With(&instance, echo.Config{
+					Service:   "server",
+					Namespace: ns,
+					Subsets:   []echo.SubsetConfig{{Annotations: echo.NewAnnotations().Set(echo.BindPodIPPorts, "6666,7777")}},
+					Pilot:     p,
+					Galley:    g,
+					Ports: []echo.Port{
+						{
+							Name:         "http5",
+							Protocol:     protocol.HTTP,
+							ServicePort:  5555,
+							InstancePort: 5555,
+						},
+						{
+							Name:         "http6",
+							Protocol:     protocol.HTTP,
+							ServicePort:  6666,
+							InstancePort: 6666,
+						},
+						{
+							Name:         "http7",
+							Protocol:     protocol.HTTP,
+							ServicePort:  7777,
+							InstancePort: 7777,
+						},
+					},
+				}).With(&client, echo.Config{
+				Service:   "client",
+				Namespace: ns,
+				Galley:    g,
+				Pilot:     p,
+				Ports: []echo.Port{
+					{
+						Name:     "grpc",
+						Protocol: protocol.GRPC,
+					},
+				},
+			}).
+				BuildOrFail(t)
+
+			clientWorkload := getWorkload(client, t)
+			serverWorkload := getWorkload(instance, t)
+			sidecar := serverWorkload.Sidecar()
+			podIP := strings.Split(sidecar.NodeID(), "~")[1]
+			expectedPortToEndpointMap := map[string]string{
+				"5555": "127.0.0.1",
+				"6666": podIP,
+				"7777": podIP,
+			}
+			configDump, err := sidecar.Config()
+			if err != nil {
+				t.Fatal(err)
+			}
+			wrapper := configdump.Wrapper{ConfigDump: configDump}
+			dcd, err := wrapper.GetClusterConfigDump()
+			if err != nil {
+				t.Fatal(err)
+			}
+			for _, cluster := range dcd.DynamicActiveClusters {
```

Comment:
If you want to introspect a config dump, do it in a unit test in my opinion. E2E tests should test behavior not implementation

Reply:
removed